services/src/main/java/org/jboss/windup/web/services/RuleDataLoader.java

```diff
@@ -165,14 +174,42 @@ public void reloadRuleData()
         }
     }
 
+    /**
+     * Clean up any stale rows (dangling rows from reloading rules).
+     */
+    private void cleanupDanglingRows()
+    {
+        this.runSqlQuiet("delete from ruleproviderentity_ruleentity where ruleproviderentity_rule_provider_entity_id " +
+                "in (select rule_provider_entity_id from ruleproviderentity where RULESPATH_RULES_PATH_ID is null);");
+        this.runSqlQuiet("delete from ruleproviderentity_technology_source where " +
+                        "rule_provider_entity_id in (select rule_provider_entity_id from ruleproviderentity where RULESPATH_RULES_PATH_ID is null);");
+        this.runSqlQuiet("delete from ruleproviderentity_technology_target " +
+                        "where rule_provider_entity_id in (select rule_provider_entity_id from ruleproviderentity where RULESPATH_RULES_PATH_ID is null);");
+        this.runSqlQuiet("delete from ruleproviderentity where RULESPATH_RULES_PATH_ID is null;");
+        this.runSqlQuiet("delete from ruleentity re where re.rule_entity_id not in " +
+                "(select rules_rule_entity_id from ruleproviderentity_ruleentity);");
```

Comment:
I'm not sure if having native SQL here is good idea. What if we change naming of tables/columns? (I think @OndraZizka  already suggested that).